[Merged by Bors] - feat(data/nat/prime): `prime.dvd_prod_iff`; golf `mem_list_primes_of_dvd_prod`

stuart-presnell opened the pull request:
Adds a generalisation of `prime.dvd_mul`: prime `p` divides the product of `L : list ℕ` iff it divides some `a ∈ L`.   The `.mp` direction of this lemma is the second part of Theorem 1.9 in Apostol (1976) Introduction to Analytic Number Theory.

Also adds the converse `prime.not_dvd_prod`, and uses `dvd_prod_iff` to golf the proof of `mem_list_primes_of_dvd_prod`.



---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

stuart-presnell committed:
feat(data/nat/prime): `dvd_prod`; golf `mem_list_primes_of_dvd_prod`

Adds a generalisation of `prime.dvd_mul`: If prime `p` divides the product of `L : list ℕ` then it divides some `a ∈ L`.   This is the second part of Theorem 1.9 in Apostol (1976) Introduction to Analytic Number Theory.

Also adds the converse `prime.not_dvd_prod`, and uses `dvd_prod` to golf the proof of `mem_list_primes_of_dvd_prod`.

stuart-presnell committed:
Upgrade `prime.dvd_prod` to an iff; fix uses elsewhere in file

stuart-presnell committed:
Fix formatting

stuart-presnell committed:
Rename lemma `prime.dvd_prod_iff`

jcommelin commented on src/data/nat/prime.lean:
```diff
@@ -492,6 +492,26 @@ theorem prime.not_dvd_mul {p m n : ℕ} (pp : prime p)
 mt pp.dvd_mul.1 $ by simp [Hm, Hn]
 
+/-- Prime `p` divides the product of `L : list ℕ` iff it divides some `a ∈ L` -/
+lemma prime.dvd_prod_iff {p : ℕ} {L : list ℕ} (pp : prime p) :
```
I guess these lemmas also hold for prime elements in more general settings (comm_rings? comm_monoid_with_zero?) I think it would make sense to reuse those proofs. Would that work for your application?

stuart-presnell replied:
Ah, I should have thought of this after @ericrbg asked for the corresponding generalisation on #10675. I've added the proof for `comm_monoid_with_zero` to a temp file for now, because I'm not sure where it best belongs. `algebra/associated` perhaps?

I'm not sure about how to use the more general lemma to prove the version for `list ℕ`, since `data/nat/prime` can't see `nat.prime_iff`.  I don't want to add an extra import (of `ring_theory.int.basic`) to such a basic file just for this one lemma.

ericrbg replied:
I do still wonder whether something like "preprime" is a good idea. This, and many similar lemmas, hold for 0 too; although, to be fair, they go further to equality, so it may still be worth special casing them. I'm not sure, it's a random idea for the futurel

stuart-presnell replied:
Would it be possible to incorporate that into the unification proposed in this thread (https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/nat.2Eprime)?

stuart-presnell replied:
I've moved the more general proof for `comm_monoid_with_zero` to `algebra/associated` and deleted the temp file. I'm still not sure how to use this to prove the version for `ℕ`, but perhaps that can be done as a separate PR after the merging of `nat.prime` and `prime` as discussed in the above Zulip thread?

ericrbg replied:
I think it's worth keeping the separate proofs (maybe making a note that it's duplicated above both for now as we're about to refactor) and we can deal with the aftermath later.

stuart-presnell replied:
I've added a TODO comment to that effect at the site of the lemma in `data/nat/prime`.

stuart-presnell committed:
Prove more general version for lists of `comm_monoid_with_zero`

stuart-presnell committed:
Temporarily moved from `algebra.associated` to new file

stuart-presnell committed:
Added docstring

stuart-presnell committed:
Moved everything to temp file

stuart-presnell committed:
Restore original proof in `nat/prime`; remove from temp file

stuart-presnell committed:
Move `prime.dvd_prod_iff` to `algebra/associated`

stuart-presnell committed:
Delete dvd_prod_iff.lean

stuart-presnell committed:
Merge branch 'master' into SP_prime_dvd_prod

stuart-presnell committed:
Add TODO comment

stuart-presnell committed:
`prod_ne_zero_of_prime` now in `algebra/big_operators/associated`

This lemma already appears in `algebra/big_operators/associated`, so deleting it from here

stuart-presnell committed:
Move `prime.dvd_prod_iff` to `algebra/big_operators/associated`
